source/iopipe/json/dom.d

```diff
@@ -7,91 +7,211 @@ import iopipe.json.parser;
 public import iopipe.json.common;
 import iopipe.traits;
 import std.traits;
+import std.conv : to;
 
 enum JSONType
 {
-    Integer,
-    Floating,
+    Integer,        // Now a hint only, stored as string
+    Floating,       // Now a hint only, stored as string
     String,
     Obj,
     Array,
     Null,
     Bool,
+    StringSSO,      // Small string optimization for strings
+    NumberSSO,      // Small string optimization for numbers
 }
 
-struct JSONValue(SType)
+// Making JSONValue a template struct with a default parameter of string
+struct JSONValue(StringType = string)
 {
-    // basically a tagged union.
+    // Tagged union
     JSONType type;
+    
+    // Small string optimization - 16 bytes should cover many common cases
+    private enum SSOSize = 16;
+    
     union
     {
-        long integer;
-        real floating;
+        // String storage (either SSO or allocated)
+        struct {
+            union {
+                char[SSOSize] sso;
+                StringType allocatedString;
+            }
+            // For SSO types, first byte can indicate length used
+            ubyte ssoLength;
+        }
+        
+        // Original value types
         JSONValue[] array;
-        JSONValue[immutable(SType)] object;
-        SType str;
+        JSONValue[StringType] object;
         bool boolean;
     }
+    
+// Get the string representation of the value
+    StringType stringForm() const
+    {
+        // You might import 'to' locally if needed, but the top-level import should suffice
+        // import std.conv : to;
+
+        with(JSONType) final switch(type)
+        {
+            case String:
+                return allocatedString;
+            case StringSSO:
+            case NumberSSO:
+                // Use std.conv.to for converting the char[] slice to StringType
+                import std.conv : to; // Add import here just to be safe/clear
+                return sso[0..ssoLength].to!StringType; // <-- Changed line
```

Comment:
Returning the slice directly is preferred, as this allocates.

I have to think about the right types for this. returning a `string` is not good, because the SSO data inside is not immutable.